### PR TITLE
[WIP] [gardening] Improvements to python-lint and check-incremental

### PR DESCRIPTION
**Pull request by alexwlchan**
#### What's in this pull request?

python-lint:
- Add more comments, in particular a top-level comment explaining the
  purpose and return code of the script.
- Add the ability to limit the linting to specific files.
- Use arithmetic comparisons when checking return codes.

check-incremental:
- Use argparse instead of a homebrew argument parser.
- Commonise the code for building and checking the md5/mtime of an
  output file.
- Use print_function for Python 3 compatibility.
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

Not fixing any Swift bugs that I know of.

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

File: utils/check-incremental [hunk 1]
```diff
@@ -8,84 +8,91 @@
...
 
-    write_obj_file = False
-    next_arg_is_output = False
-    compare_time = True
-    output_file = None
+    parser.add_argument('-o', '--output', metavar='OUTPUT_FILE', required=True,
+                        help='output file to compare')
+    parser.add_argument('-c', required=True,
+                        action='store_true', dest='write_obj_file',
+                        help='write_obj_file')  # @@AWLC: wut?
```
Comment (alexwlchan) [inline utils/check-incremental]: I wasn’t really sure what this argument was supposed to be for – the original code exits if this flag is missing, but doesn’t do anything else with it.

I’m guessing this is a flag that gets passed to the Swift compiler somewhere?

File: utils/check-incremental [hunk 2]
```diff
@@ -8,84 +8,89 @@
...
-            write_obj_file = True
-        elif arg == '-disable-incremental-llvm-codegen':
-            compare_time = False
-        elif arg == '-o':
-            next_arg_is_output = True
+    parser.add_argument('-o', metavar='OUTPUT_FILE', required=True,
+                        help='output file to compare')
+    parser.add_argument('-c', required=True,
+                        action='store_true', dest='write_obj_file',
+                        help='write_obj_file')  # @@AWLC: wut?
```
Comment (alexwlchan) [inline utils/check-incremental]: I wasn’t really sure what this argument was supposed to be for – the original code exits if this flag is missing, but doesn’t do anything else with it.

I’m guessing this is a flag that gets passed to the Swift compiler somewhere?

**Comment by alexwlchan** [comment]
The python-lint changes are entirely selfish – when I’m making edits, it’s easier to lint if I can restrict the checking to files that I’ve modified. It’s not like the original linter is slow – less than 10&nbsp;seconds on my Mac – but making it even faster never hurts.

File: utils/python-lint [hunk 3]
```diff
@@ -1,16 +1,26 @@
 #!/bin/bash
+# This script is a linter for Python files in the Swift codebase.
+# If there are any linting errors, the script exits with code 1; otherwise
+# it exists with retcode 0.
+#
+# Optionally, you can supply the name of one or more files (relative to the
+# root of the codebase) to limit the linting to those files.
```
Comment (gottesmm) [inline utils/python-lint]: Can you add a note here that this is the script used by @swift-ci please Python lint.

That is the actual purpose of this script: to make it really easy to reproduce how @swift-ci lints so one can check for failures on ones own machine.

File: utils/python-lint [hunk 4]
```diff
@@ -1,16 +1,26 @@
...
+if (( $? != 0 )); then
   echo "Missing modules flake8 or flake8-import-order. Please be sure to install these python packages before linting."
   exit 1
 fi
 
+# This script lives in the utils directory in the Swift codebase, so
+# this command should take us to the root of the codebase.
 cd "$SCRIPT_DIR/.."
-if [[ $? != 0 ]]; then
+if (( $? != 0 )); then
```
Comment (gottesmm) [inline utils/python-lint]: Are you trying to use the arithmetic extensions to bash here? Why not just do:

[[ $? -ne 0 ]]

**Comment by alexwlchan** [comment]
(Lightbulb moment)

If all these arguments are getting passed along to a build step somewhere, we don’t want argparse in the mix, because it chokes on all the extra arguments. Of course that only occurs to me after opening the PR. I’ll rethink and come back with a new PR.

File: utils/python-lint [hunk 5]
```diff
@@ -1,16 +1,26 @@
...
+#
+# Optionally, you can supply the name of one or more files (relative to the
+# root of the codebase) to limit the linting to those files.
 
+# Full directory name of this script.
 SCRIPT_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
 
 python -c 'import flake8; import flake8_import_order' 2>/dev/null
-if [[ $? != 0 ]]; then
+if (( $? != 0 )); then
```
Comment (gottesmm) [inline utils/python-lint]: Are you trying to use the arithmetic extensions to bash here? Why not just do:

[[ $? -ne 0 ]]

**Comment by gribozavr** [comment]
@swift-ci Please Python lint

**Pull request closed without merging**